docs/proposals/compute-resource-metrics-api.md

```diff
@@ -0,0 +1,179 @@
+
+```
+$ kubectl top kubernetes-minion-abcd
+Pod: monitoring-heapster-abcde
+CPU:
+    mean: 0.12 cores
+    95th: 1.10 cores
+    max:  1.91 cores
+Memory:
+    mean: 300MB (10%)
+    95th: 330MB (11%)
+    max:  395MB (13%)
+```
+
+A second user will be the scheduler. To assign pods to nodes efficiently, the
+scheduler needs to know the current free resources on each node.
+
+## Proposed endpoints
+
+	/api/v1/namespaces/myns/podMetrics/mypod/[derived|raw]
+	/api/v1/nodeMetrics/myNode/[derived|raw] 
+
+The derived metrics include the mean, max and a few percentiles of the list of
+values.
+
+The raw metrics are the values themselves, thus the data will be much heavier.
+
+## Data gathering
+
+We can either use a Push or a Poll based system:
+
+* Push - each node would periodically POST their derived and raw metrics to
+  the API server. Kubectl and the schedule would watch the API server to get
+  the latest metrics.
+
+* Poll - heapster would periodically poll a kubelet API endpoint on each node
+  to fetch its node and pods metrics. Then Heapster will have to implement the
+  APIs mentioned above.
+
+Our initial plan is to follow the push based system.
+
+The API server will not hold any of this data directly. For our initial
+purposes, it will hold the most recent metrics obtained from each node in
+etcd. Then, when polled for metrics, the API server would only serve said most
+recent data per node. The clients could cache this data if they needed to keep
+a historical record of it, for example Heapster.
+
+In the future, we might want to store the metrics differently:
+
+* via heapster - Since heapster keeps data for a period of time, we could
+  redirect requests to the API server to heapster instead of using etcd. This
+  would also allow serving metrics other than the latest ones.
+
+## Data structure
+
+```Go
+type DerivedPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]ResouresMetrics
+}
+
+type DerivedNodeMetrics struct {
+	TypeMeta
+	ObjectMeta // should have node name
+	NodeMetrics      ResourcesMetrics
+	SystemContainers map[string]ResourcesMetrics
+}
+
+type ResourcesMetrics map[resource.Type]DerivedMetrics
+
+type DerivedMetrics struct {
+	// Time at which these metrics were gathered
+	Time util.Time `json:"timestamp"`
+
+	// 1m, 1h and 1d as a start
+	Metrics map[time.Duration]Metrics
+}
+
+type Metrics struct {
+	// Average over the collected sample.
+	Mean resource.Quantity `json:"mean"`
+	// Max seen over the collected sample.
+	Max resource.Quantity `json:"max"`
+	// 95th percentile over the collected sample.
+	NinetyFive resource.Quantity `json:"95th"`
+}
+
+type RawPodMetrics struct {
+	TypeMeta
+	ObjectMeta // should have pod name
+	// the key is the container name
+	Containers map[string]struct {
+		ContainerReference *Container
+		RawMetrics         []RawMetric
+	}
+}
+
+type RawNodeMetrics struct {
```

Comment:
But that's provided by Prometheus at /metrics - this would transform that into a structured form and send it to the apiserver (for push) or be exposed as a new endpoint on the kubelet (for pull)?

The transformation seems somewhat lossy, although we do need a structured form.  Would a pull model in the short term have to individually query each pod?  Why not make those retrievable all at once?  Does the scheduler need to get pod metrics for the pods already scheduled on a node at all, or can it use only the free metrics?